src/openrct2/ride/Vehicle.cpp

```diff
@@ -7280,26 +7280,41 @@ void Vehicle::UpdateAnimationAnimalFlying()
 void Vehicle::UpdateAdditionalAnimation()
 {
     uint8_t targetFrame{};
-    uint8_t curFrame{};
-    uint32_t eax{};
-
     auto vehicleEntry = Entry();
     if (vehicleEntry == nullptr)
     {
         return;
     }
+    // similar animation styles all update the same
     switch (vehicleEntry->animation)
     {
-        case VEHICLE_ENTRY_ANIMATION_MINITURE_RAILWAY_LOCOMOTIVE: // loc_6D652B
+        case VEHICLE_ENTRY_ANIMATION_MINITURE_RAILWAY_LOCOMOTIVE:
+        case VEHICLE_ENTRY_ANIMATION_SWAN:
+        case VEHICLE_ENTRY_ANIMATION_CANOES:
+        case VEHICLE_ENTRY_ANIMATION_ROW_BOATS:
+        case VEHICLE_ENTRY_ANIMATION_WATER_TRICYCLES:
+        case VEHICLE_ENTRY_ANIMATION_HELICARS:
             animationState += _vehicleVelocityF64E08;
-            targetFrame = (animationState >> 20) & 3;
+            // Step 1: shift the value right by 8 less than you want
+            // Step 2: mask by 0xFF to get only 8 bits of value
+            // Step 3: multiply by the number of frames
+            // Step 4: shift right by 8 to remove the masked bits
+            // Step 5: you are left with the frame number
+            targetFrame = ((animationState >> std::clamp(vehicleEntry->AnimationSpeed - 8, 0, 32) & 0xFF)
```

Comment:
This line is not clear if you want to have
```cpp
(animationState >> std::clamp(vehicleEntry->AnimationSpeed - 8, 0, 32)) & 0xFF
```
or
```cpp
animationState >> (std::clamp(vehicleEntry->AnimationSpeed - 8, 0, 32) & 0xFF)
```
Make this explicit.

Reply:
Also there are magic values like 8 and 32, they should be constants.

Reply:
> This line is not clear if you want to have
> 
> ```c++
> (animationState >> std::clamp(vehicleEntry->AnimationSpeed - 8, 0, 32)) & 0xFF
> ```
> 
> or
> 
> ```c++
> animationState >> (std::clamp(vehicleEntry->AnimationSpeed - 8, 0, 32) & 0xFF)
> ```
> 
> Make this explicit.

The second option is the same as not having `& 0xFF` because the value is already lower than 256.

Reply:
> Also there are magic values like 8 and 32, they should be constants.

They are the sizes of integer values, but I can do that.

Reply:
> > This line is not clear if you want to have
> > ```c++
> > (animationState >> std::clamp(vehicleEntry->AnimationSpeed - 8, 0, 32)) & 0xFF
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or
> > ```c++
> > animationState >> (std::clamp(vehicleEntry->AnimationSpeed - 8, 0, 32) & 0xFF)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Make this explicit.
> 
> The second option is the same as not having `& 0xFF` because the value is already lower than 256.

That is why I'm saying use parenthesis to make your intent clear here.

Reply:
I already use parenthesis so I'm not sure why you are saying this.

Reply:
Because who ever reads the code may question if want to shift with a masked value of 0xFF or if you want to mask the resulting value with 0xFF. There is a lot going on with this single line already, but also as suggested its best to move this into a function and it might be best to use a few variables to make this even easier to read.